### PR TITLE
feat: implement local ML stem separation with chunking

### LINKED ISSUE #106: Epic: Spectral Analysis & Source Separation (Stems)
## Description
Replace the mock role extraction by integrating a local-first source separation model to split the ingested audio into discrete instrumental/vocal stems.

## Tasks
- [ ] Evaluate and integrate a lightweight local separation model (e.g., Demucs, Spleeter) compatible with CPU/MPS/CUDA.
- [ ] Implement inference chunking to prevent Out-Of-Memory (OOM) errors on standard user machines.
- [ ] Map separated stems to BandScope's internal `RoleType` models (Vocals, Bass, Drums, Other/Chords).
- [ ] Implement secure model weight downloading and supply chain verification according to `dependency-policy.md`.

## Definition of Done
The analysis engine successfully splits a given track into standard stems locally without network dependencies for inference, and the stems are available to downstream analysis pipelines.

### PR DESCRIPTION
**Pull request by seonghobae**
## Description

This PR integrates local-first source separation into the `bandscope-analysis` engine. It utilizes `torch` and `torchaudio` (demucs) to split ingested audio into discrete instrumental/vocal stems. 

**Key Changes:**
- `AudioStemSeparator` class with device management (CUDA/MPS/CPU).
- Chunked inference logic to prevent Out-Of-Memory (OOM) errors on devices with limited memory.
- Integration with `cli.py` to support the `--separate-stems` flag.
- Supply chain updates to include PyTorch dependencies.
- 100% test coverage maintained via extensive mocks in `test_audio_separator.py` and `test_cli.py`.

Resolves #106

## Security Notes
- **Untrusted Inputs:** Audio models are loaded locally without external unauthenticated network requests after the initial weight download (controlled via Torch Hub).
- **Subprocesses:** N/A (Runs within the same Python process).
- **Memory Management:** Memory bounds are explicitly maintained through chunked processing and garbage collection via `gc.collect()` and clearing cache.

### CODE REVIEW

**Comment by seonghobae** [comment]
@coderabbitai review

**Comment by seonghobae** [comment]
@coderabbitai review

**Comment by seonghobae** [comment]
@coderabbitai resolve

**Comment by seonghobae** [comment]
@coderabbitai resolve

**Comment by seonghobae** [comment]
@coderabbitai resolve

**Comment by seonghobae** [comment]
Updated #111 to fix the failing checks after the scheduler/develop merge.

What changed:
- Restored scheduler artifacts from current `develop` so this PR no longer deletes them.
- Resolved the Python dependency set for stem separation as a compatible Demucs/Torch/Torchaudio stack: `numpy<2.0`, `torch<=2.2.2`, `torchaudio<=2.2.2`, while keeping `urllib3>=2.7.0` and Pygments 2.20.0 in the lockfile.
- Avoided importing real Demucs/Torchaudio in unit tests by injecting fake `demucs.*` modules.
- Added coverage for missing model weights, checksum mismatch, invalid `--separate-stems` input, and temporal-probe failure.

Validation:
- `python3 scripts/checks/verify_supply_chain.py`
- `uv run ruff check src tests`
- `uv run ruff format --check src tests`
- `uv run mypy src`
- `uv run pytest tests --cov=src/bandscope_analysis --cov-report=term-missing --cov-fail-under=100` (282 passed, 100% coverage)
- `./scripts/harness/quickcheck.sh` (passed after documented `npm install`; npm warned local Node is v24 while repo asks for >=22 <23)
- `git diff --check`

Security Notes:
- Stem-separation model weights remain pre-provisioned local artifacts; no runtime model download path was added.
- Local model loading now has explicit tests for missing weights and checksum mismatch fail-closed behavior.
- Dependency constraints avoid the broken NumPy 2.x/Torch/Torchaudio combination seen in CI while keeping the vulnerability-gated lock entries current.

**Comment by seonghobae** [comment]
@coderabbitai approve

**Comment by seonghobae** [comment]
@coderabbitai review

**Comment by seonghobae** [comment]
Correction to the previous validation note: the final pushed dependency stack is `torch>=2.11.0,<2.12.0` and `torchaudio>=2.11.0,<2.12.0`, not Torch 2.2.x. I changed this after dependency-review and Trivy flagged GHSA-53q9-r3pm-6pq6 against Torch 2.2.2.

Additional validation after the final lockfile:
- `uv run --project services/analysis-engine python -c "import torch, torchaudio; print(torch.__version__); print(torchaudio.__version__)"` -> 2.11.0 / 2.11.0
- `uv sync --project services/analysis-engine --group dev --frozen`
- `uv run --project services/analysis-engine --with pip-audit==2.8.0 pip-audit --local --strict` -> no known vulnerabilities
- `./scripts/harness/quickcheck.sh` -> passed

Security Notes update:
- The vulnerable Torch 2.2.2 lock entry was removed.
- Torch and Torchaudio are pinned to the same patched minor line to avoid ABI mismatch while satisfying the PyTorch advisory patched-version requirement.

**Comment by seonghobae** [comment]
@coderabbitai review

**Comment by seonghobae** [comment]
@coderabbitai approve

**Comment by seonghobae** [comment]
Closing this PR instead of merging it.

Current live state: #111 is approved but `DIRTY`, and its branch diff against `develop` is no longer a narrow stem-separation change. It touches 55 files, including workflow/action-pin files, frontend package/test files, section utilities, YouTube/security tests, supply-chain verification tests, and several analysis modules that have already been updated by newer merged PRs.

Concrete blockers:
- The branch would roll back current `develop` work, including `.github/workflows/*` pin/hardening changes, `scripts/checks/verify_supply_chain.py`, `services/analysis-engine/src/bandscope_analysis/sections/utils.py`, YouTube tests/security behavior, GrooveMap updates, and other merged analysis/test changes.
- It adds new direct runtime dependencies `demucs`, `torch`, and `torchaudio` with a very large transitive footprint, while also downgrading the current `numba` constraint. That needs current dependency-admission evidence under the BandScope dependency policy before it can ship.
- The model artifact inventory only records an 8-character checksum prefix (`f7e0c4bc`) for `htdemucs` and points at `~/.cache/torch/hub/checkpoints/`. Model artifacts need full provenance/checksum and release inventory evidence, not a user-home cache path plus a prefix check.
- The feature crosses the local model/file-processing trust boundary, so it needs a fresh app-security review for untrusted audio input, model artifact loading, safe failure, logging/privacy impact, and tests.
- The PR is stale enough that safely merging would require rebuilding the change on top of current `develop`, not resolving conflicts in place.

Please reopen as a fresh, focused PR if stem separation is still desired: start from current `develop`, keep the dependency/model admission in the same PR, use full artifact checksums, and avoid rolling back unrelated security/workflow/test changes.
